Added Pre-Upgrade report command to use relevant daemon endpoint

artmello opened the pull request.

artmello committed:
Added check-system command to execute a group of checks on target system

artmello commented:
Depending on:
https://github.com/leapp-to/leapp-actors/pull/19
https://github.com/leapp-to/leapp-go/pull/29

bocekm commented:
@artmello, just a note about naming. We should think about the report in more generic way, i.e. the report won't be used only for the pre-upgrade phase, but after all also for the pre-migration and pre-containerization phases. We talked about it with @bertinatto and @mfranczy and they said they would need to provide information to the user too before the actual containerization phase.

artmello commented:
@bocekm following up after a couple of meetings, I changed this PR to instead of calling a "pre upgrade report" endpoint, it's now adding a new command to leappctl name "check-system" and this command has a "checks" param. Via such parameter user can specify which group will be executed on the target machine. Some sample usage:
$leappctl check-system --checks pre-upgrade
$leappctl check-system --checks pre-migration
$leappctl check-system --checks some_other_check_actor

Hopefully this can address current and some future necessities. What do you think?

bocekm commented:
So far, I've been using just snactor to run actors. I'll try the leappctl on Monday.

artmello committed:
Added support for getting HTML result from checks

- Added --html option to check-system so user can get an HTML output from checks results
- Added --out option to check-system so user can decide to save result into a file

shaded-enmity commented on leappctl/cmd/check_system.py:
```diff
@@ -0,0 +1,63 @@
+            print(html_content)
+
+    else:
+        json_pprint = json.dumps(resp_body, sort_keys=True, indent=4, separators=(',', ': '))
```
I'd not pretty print the JSON here (or in any actor), while it may be nice for debugging it just wastes cpu cycles and uses much more space.

shaded-enmity commented on leappctl/cmd/check_system.py:
```diff
@@ -0,0 +1,63 @@
+        json_pprint = json.dumps(resp_body, sort_keys=True, indent=4, separators=(',', ': '))
+        if output_file:
+            with open(output_file, 'w+') as f:
+                f.write('Response:\n{0}\n'.format(json_pprint))
```
Why not just write the data directly (ditto below), i.e. without the 'Response:' - If I wanted to use this as a plumbing command I'd have to strip first line manually just to get to the real JSON data.

artmello committed:
Changed check_system subcommand output

- Do not pretty print or format output anymore

bertinatto commented on leappctl/cmd/check_system.py:
```diff
@@ -0,0 +1,58 @@
+              default=False,
+              help='Display check result as HTML')
+@click.option('--out',
+              '-o',
```
You could use Click's `click.File` as `type` here, so it would take care about opening it later.

bertinatto commented on leappctl/cmd/check_system.py:
```diff
@@ -0,0 +1,58 @@
+    resp = post(CMD, req_body)
+
+    # Pretty-print response
+    resp_body = resp.json()
```
TODO for our future selves:

1) handle errors here;
2) decide how we want to print results in `leappctl`; raw json as we are doing or a more user-friendlier way?

vinzenz replied:
@bertinatto leappctl should actually not be printing JSON eventually (maybe only on request (something like a --json parameter)) 
It's supposed to be a TUI

bertinatto replied:
+1

bertinatto commented on leappctl/cmd/check_system.py:
```diff
@@ -0,0 +1,58 @@
+            with open(output_file, 'w+') as f:
+                f.write(json.dumps(resp_body))
+        else:
+            print(json.dumps(resp_body))
```
I'd rather do something like (if possible):

```
if display_html:
    resp_html = post(HTML_OUT_CMD, resp_body['data'])
    resp_html_body = resp_html.json()
    try:
        content = resp_html_body['data']['html_output'][0]['value']
    except LookupError:
        (...)
else:
    content = json.dumps(resp_body)

if output_file:  # assuming this is a click.File
    output_file.write(content)
else:
    print(content)
```

artmello committed:
Improved check_system subcommand code

artmello commented:
@bertinatto thanks for reviewing, hopefully it will look better now

vinzenz reviewed (approved):
LGTM

bertinatto commented:
Thanks, @artmello. LGTM.
